### ISSUE TITLE
Ready event?

### ISSUE THREAD

**Issue by endash** [gss/engine#169]
In my efforts to integrate GSS and Ember.js it was quickly apparent that there is a slight delay in the application of GSS rules on the *initial* load of the app. Subsequent changes are seamless, as far as I can tell. I tried using `solved` to delay initialization of the app until after GSS was 'up and running,' but for whatever reason (possibly a race condition re: the loading/processing of the linked GSS stylesheet) about 10% of the time a `solved` event isn't kicked off and the app, stuck waiting on a solved event that's never coming, sits waiting, thus not triggering any DOM changes that might then trigger GSS to solve.

From what I can tell there's no 1 event that is guaranteed to be fired after a solution cycle.

**Comment by Inviz** [comment]
GSS is initialized ondomready. It also tries to re-draw on readystatechange/load events to account for loaded fonts and images. You can probably force GSS to initialize stylesheets by calling engine.compile() before `</body>` or somewhere around rendering your templates. 

`solved` event is supposed to be called after every full resolving tick once at the very end. 

It should not load stylesheets before it's ready to handle it, so if there is a race condition/bug it might be in some way how async files are handled and sorted. 

External stylesheet are fairly undertested. There were atleast two other related bugs fixed since 2.0 release on 2.1.x branch. I'd like to have your case covered, can you provide more details, or even better - a page with reproducable problem?

**Comment by endash** [comment]
I guess this could be two issues: (1) There's no guaranteed-to-be-called-at-the-end-of-solve event to listen to, just as a matter of fact, not necessarily making any judgements as to whether such an event would be useful to others (2) There's a potential race condition involving a combination of external stylesheets and Ember.JS initializers. 

The first is straightforward and either there's a case for having a guaranteed event or not. I found that I was able to rely on 'flush', but there might be other use cases that wouldn't be able to. I don't want to speculate on that. 

The second is so specific and ultimately so complex that it may very well be fruitless, and would at least be a lot of work, to try to track down the cause one way or another. 

I don't expect the race condition thing to be solved, this was more of a "what event can I use that is guaranteed to get called at least once?" In my case one of the earlier events sufficed.

**Comment by endash** [comment]
For lack of desire to dig through and figure out exactly what the timing issues are, I'm gonna close, unless the minor event thing is worth talking about.

**Comment by Inviz** [comment]
`solved` is guaranteed to be called at the end of the solution, if the tick had any side effects (new elements added that match css rules/constraints). It's something like "onchange". If nothing was detected to be changed, `solved` is not called. If you enable logging via `?log=1` parameters in the URL, you can see in console if there were changes or not, both objects next to the tick group will be empty.

 'flush' may be called multiple times per one solution, so keep that in mind. 

 I think i could add `finish` event that would be fired even if nothing has changed. If your problem was not some engine deadlock due to async stylesheets, it would help you. I'll close issue when event is added.

Thanks!

**Comment by Inviz** [comment]
![screen shot 2015-02-17 at 2 18 32 pm](https://cloud.githubusercontent.com/assets/2708/6238437/e4dec8c2-b6af-11e4-922c-bffc4663eacd.png)

**Comment by endash** [comment]
Jeez I just doublechecked and the method with all the events firing is `commit`, not `solve`. My bad.

As part of my testing to see how I could use GSS events to delay the Ember initialization I added a 'finish' event myself to have a guaranteed 'final event,' but for production went with `once` on `flush` because that seemed to work (it was never not called, and seemed to get called late enough that there was no "flicker" in the App UI).

What I didn't try was to add an event at the *start* to see if I even needed to wait that long. It would be *nice* to know exactly what parts of GSS need to be run to prevent the UI flicker, but from a pragmatic standpoint I don't think it would make much difference to me, so long as I could depend on any single event being consistently fired late enough.

(Just for clarification on exactly what I'm doing: Ember.js allows initializers to halt app booting, primarily to handle any asynchronous setup. I am halting the app boot, and in my callback for the event (in this case, now it's `flush`) I was resuming app boot. My initial attempt of `solved` would occasionally not fire due to that race condition, and the boot would stall forever.)

**Comment by Inviz** [comment]
I think using external stylesheets will cause flicker unless you hide the whole thing via css like `html:not(.gss-ready) { opacity: 0 }`. Inline stylesheets will be executed synchronously, but still only after domready.

Thanks for clarification. Hope you can confirm later if `finish` helps (needs to be added to two places in `commit` method) or was it some more "real" issue.

Please continue with feedback, questions, bugs and stuff. Happy to help.

**Comment by endash** [comment]
The CSS solution is a good one, but I'd rather trap the delay programmatically, where it can be measured, otherwise handled if needed, etc.  The insight that the external stylesheet is behind the flickering is helpful and I'll keep that in mind if/as I explore further. Now I get to actually *use* GSS, so I'm happy right now :)

Oh, hey, since the sausage making is entirely hidden from my users by Ember, I could in theory load things in a way that wouldn't otherwise be user friendly. Such as, potentially, a pre-compiled script file with AST. Would there be (more than trivial) performance benefits to doing that, do you think, and would that be a path to resolving the async flicker?

**Comment by Inviz** [comment]
There will be no flicker if GSS inline stylesheets are applied in the same tick as templates are rendered/before it. Use `<script type="text/gss-ast">` with output of `GSS.Parser.parse(...).commands`.

Yes, there will be noticable performance benefit, as parser is not very optimized due to being auto-generated from PEGs. We'll rewrite it one day to be faster.

**Comment by endash** [comment]
Out of curiosity, are there benchmarks anywhere? Not looking for sub-millisecond precision, just order of magnitude.

**Comment by Inviz** [comment]
on big pages like thegrid.io with a lot of stylesheets, we get about 15-25% initial boot time spent in parser

**Comment by endash** [comment]
I'm less concerned about boot time as that's a one-time hit. What would be interesting, and particularly relevant re: mobile, are update times, and how they scale with ruleset size, and dom size.

**Comment by endash** [comment]
Sorry I'm mixing issues here :P

**Comment by Inviz** [comment]
Fair enough, there're no benchmarks like that yet. DOM/rulesets doesn't matter in itself, however constraint graphs make a big deal. 

There's a world of difference between a cassowary instance with 500 constraints vs 50 constraints. Check http://gridstylesheets.org/?log=1 it shows that we separated graphs to be relatively small. So it resizes quite passibly, even if we didn't optimize too hard for speed yet on that website. Try resizing the window on my machine it's pretty fast. Console shows timings as well when ?log=1 is enabled. We're aiming for 60fps scrolling and 30fps resizing on modern machines. Not always quite there yet.

To separate constraint graphs, and thus harnessing multiple cassowaries for speed, you have make sure that you don't use linear variables processed by cassowary. Things like grid size, configurable breakpoint and other "constants" should be passed as suggested values. They don't join graphs and can be updated efficiently. We're adding assignments in 2.1.x that allow people to use suggested values from within gss stylesheets.

The biggest performance offender is if block *with* else part, because it makes cassowary drop the iterative state to recompute everything from scratch. We'll have better performance for elseif blocks in 2.2 (but we didnt release 2.1 yet)

So in the order of impact, the steps to improve page performance would be:

1) Make sure there're as many constraint graphs as possible. Check out `position: relative` section in docs, it shows how to separate big rectangles from being computed by the same engine. Basically, it relies on CSS to make big portions of the website "flow", while only positioning content within rectangles.

2) Don't use if/else blocks. If a tick only adds or only removes constraints from a graph, it will re-use state. If it replaces constraint in a graph, it has to drop the state. 

3) Pre-parse stylesheets to avoid initial performance hit.

**Comment by endash** [comment]
So an individual constraint graph is like an isolated island, that can be resolved independent of the rest of the constraints? That makes a lot of sense. I think that's going to get ugly fast for me when I start implementing my native-esque UI framework on top of GSS, and there *aren't* any islands. But that's ok my project is still mostly experimental and I'll cross those performance bridges when I come to them.

I'll definitely keep the `else` thing in mind, that's very good to know, thanks!

**Comment by endash** [comment]
Along those lines... if two otherwise isolated sets of constraints both reference ::window, do they count as linked or does the immutability of window break what might otherwise be a connection?

**Comment by Inviz** [comment]
No. window size, scroll position, intrinsic/computed values and suggested variables all don't join graphs and are optimized for change via cassowary edit variables.

Thanks for good questions.

**Comment by endash** [comment]
Thank you for taking the time :) Much appreciated.
